src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -423,17 +377,108 @@ private async Task ExecuteRequests(Hl7.Fhir.Model.Bundle responseBundle, HTTPVer
+                        Api.Resources.Forbidden);
+                }
+            }
+
+            return entryComponent;
+        }
+
+        private void SetupContexts(RouteContext request, HttpContext httpContext)
+        {
+            IFhirRequestContext originalFhirRequestContext = _fhirRequestContextAccessor.FhirRequestContext;
+
+            request.RouteData.Values.TryGetValue("controller", out object controllerName);
+            request.RouteData.Values.TryGetValue("action", out object actionName);
+            request.RouteData.Values.TryGetValue(KnownActionParameterNames.ResourceType, out object resourceType);
+            var newFhirRequestContext = new FhirRequestContext(
+                httpContext.Request.Method,
+                httpContext.Request.GetDisplayUrl(),
+                originalFhirRequestContext.BaseUri.OriginalString,
+                originalFhirRequestContext.CorrelationId,
+                httpContext.Request.Headers,
+                httpContext.Response.Headers)
+            {
+                Principal = originalFhirRequestContext.Principal,
+                ResourceType = resourceType?.ToString(),
+                AuditEventType = _auditEventTypeMapping.GetAuditEventType(
+                    controllerName?.ToString(),
+                    actionName?.ToString()),
+            };
+
+            _fhirRequestContextAccessor.FhirRequestContext = newFhirRequestContext;
+
+            _bundleHttpContextAccessor.HttpContext = httpContext;
+        }
+
         private void PopulateReferenceIdDictionary(IEnumerable<EntryComponent> bundleEntries, IDictionary<string, (string resourceId, string resourceType)> idDictionary)
         {
             foreach (EntryComponent entry in bundleEntries)
             {
+                if (entry.Request.Method == HTTPVerb.PUT && entry.Request.Url.Contains('?', StringComparison.InvariantCultureIgnoreCase) && !string.IsNullOrEmpty(entry.FullUrl))
+                {
+                    if (!idDictionary.ContainsKey(entry.FullUrl))
+                    {
+                        // This id is new to us
+                        var insertId = _resourceIdProvider.Create();
+                        entry.Resource.Id = insertId;
+
+                        idDictionary.Add(entry.FullUrl, (insertId, entry.Resource.TypeName));
+                    }
+
+                    continue;
+                }
+
                 if (entry.Request.Method != HTTPVerb.POST)
                 {
                     continue;
                 }
 
-                // We've already come across this ID
-                if (!string.IsNullOrWhiteSpace(entry.FullUrl) && !idDictionary.ContainsKey(entry.FullUrl))
+                if (string.IsNullOrWhiteSpace(entry.FullUrl))
+                {
+                    continue;
+                }
+
+                // We've not come across this ID
+                if (!idDictionary.ContainsKey(entry.FullUrl))
```

Comment:
the logic below (484-487) is now identical to logic above (461-464). Can we modify the condition to simplify the logic and add some comments on the details?